turbopack/crates/turbo-tasks-malloc/src/counter.rs

```diff
@@ -1,124 +1,198 @@
-const KB: usize = 1024;
-/// When global counter is updates we will keep a thread-local buffer of this
-/// size.
-const TARGET_BUFFER: usize = 100 * KB;
-/// When the thread-local buffer would exceed this size, we will update the
-/// global counter.
-const MAX_BUFFER: usize = 200 * KB;
+lazy_static::lazy_static! {
+    static ref ACTIVE: Mutex<GlobalData> = Mutex::new(Default::default());
+}
 
 #[derive(Default)]
+struct GlobalData {
+    /// Data accumulated from retired threads
+    allocations: usize,
+    deallocations: usize,
+    allocation_count: usize,
+    deallocation_count: usize,
+    active: FxHashSet<RegisteredCounters>,
+}
+
+struct AtomicCounters {
+    allocations: AtomicUsize,
+    deallocations: AtomicUsize,
+    allocation_count: AtomicUsize,
+    deallocation_count: AtomicUsize,
+}
+
+impl AtomicCounters {
+    const fn new() -> Self {
+        Self {
+            allocations: AtomicUsize::new(0),
+            deallocations: AtomicUsize::new(0),
+            allocation_count: AtomicUsize::new(0),
+            deallocation_count: AtomicUsize::new(0),
+        }
+    }
+
+    fn from_allocation_counters(counter: AllocationCounters) -> Self {
+        Self {
+            allocations: AtomicUsize::new(counter.allocations),
+            deallocations: AtomicUsize::new(counter.deallocations),
+            allocation_count: AtomicUsize::new(counter.allocation_count),
+            deallocation_count: AtomicUsize::new(counter.deallocation_count),
+        }
+    }
+    fn to_allocation_counters(&self) -> AllocationCounters {
+        AllocationCounters {
+            allocations: self.allocations.load(Ordering::Relaxed),
+            deallocations: self.deallocations.load(Ordering::Relaxed),
+            allocation_count: self.allocation_count.load(Ordering::Relaxed),
+            deallocation_count: self.deallocation_count.load(Ordering::Relaxed),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Hash, Copy, Clone)]
+struct RegisteredCounters(*const AtomicCounters);
+unsafe impl Send for RegisteredCounters {}
+unsafe impl Sync for RegisteredCounters {}
+
 struct ThreadLocalCounter {
-    /// Thread-local buffer of allocated bytes that have been added to the
-    /// global counter desprite not being allocated yet. It is unsigned so that
-    /// means the global counter is always equal or greater than the real
-    /// value.
-    buffer: usize,
-    allocation_counters: AllocationCounters,
+    counters: AtomicCounters,
+    registered: bool,
 }
 
 impl ThreadLocalCounter {
     const fn new() -> Self {
         Self {
-            buffer: 0,
-            allocation_counters: AllocationCounters::new(),
+            registered: false,
+            counters: AtomicCounters::new(),
         }
     }
+    #[inline]
+    #[cold]
+    fn register(&mut self) {
+        let ptr = RegisteredCounters(&self.counters as *const AtomicCounters);
+        let inserted = ACTIVE.lock().unwrap().active.insert(ptr);
```

Comment:
This can cause a memory allocation.

Reply:
indeed.  discovered as soon as i added a mulithreaded test. it crashes entering the `unsafe-cell` the second time.

this means we are hosed with this approach.

furthermore rust disallows allocators from accessing thread local variables with drop implementations on some platforms so that also doesn't work.

hello async-safety requirements.

So this means i need to

1. preallocate a global array of AtomicAlllocationCounters
2. allocate an index to each thread on first access (could literally be a random number)
3. send increments to both places

This will eliminate the need for `drop` and the need for `unload` at the expense of needing to perform two sets of increments.  Still by using an array we can statistically eliminate cross thread contention